sop-extras/src/Data/SOP/Index.hs

```diff
@@ -61,34 +59,31 @@ pattern IS ::
   => xs ~ (x' ': xs')
   => Index xs' x
   -> Index xs x
-pattern IS z <- Index (S (Index -> z)) where
-  IS (Index z) = Index (S z)
+pattern IS z <- Index (S (Index -> !z)) where
+  IS (Index !z) = Index (S z)
 
 {-# COMPLETE IZ, IS #-}
 
 indices :: forall xs. SListI xs => NP (Index xs) xs
 indices = case sList @xs of
     SNil  -> Nil
-    SCons -> IZ :* hmap (Index . S . getIndex) indices
+    SCons -> IZ :* hmap IS indices
 
-dictIndexAll :: All c xs => Proxy c -> Index xs x -> Dict c x
-dictIndexAll p = \case
-    IZ     -> Dict
-    IS idx -> dictIndexAll p idx
+dictIndexAll :: forall c xs x. All c xs => Proxy c -> Index xs x -> Dict c x
+dictIndexAll p (Index idx) = hcollapse $ hcmap p (\Refl -> K Dict) idx
 
-injectNS :: forall f x xs. Index xs x -> f x -> NS f xs
-injectNS idx x = case idx of
-    IZ      -> Z x
-    IS idx' -> S (injectNS idx' x)
+injectNS :: forall f x xs. All Top xs => Index xs x -> f x -> NS f xs
```

Comment:
You can avoid the `All Top` constraint (same as `SList`) here and below by using `npToSListI` on `idx`, such that it doesn't need to be propagated.